test/integration/apimachinery/watchlist_test.go

```diff
@@ -0,0 +1,204 @@
+*/
+
+package apimachinery
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"testing"
+	"time"
+
+	"github.com/spf13/pflag"
+	"github.com/stretchr/testify/require"
+
+	"github.com/google/go-cmp/cmp"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+	genericapiserveroptions "k8s.io/apiserver/pkg/server/options"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/cmd/kube-apiserver/app/options"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+func TestReflectorWatchListFallback(t *testing.T) {
+	ctx := context.TODO()
+
+	t.Log("Starting etcd that will be used by two different instances of kube-apiserver")
+	etcdOptions, etcdTearDownFn := startEtcd(t)
+	defer etcdTearDownFn()
+
+	t.Log("Starting the first server with the WatchList feature enabled")
+	clientSet, apiServerTearDownFn := startServerWatchListEnabled(t, ctx, etcdOptions)
+	defer apiServerTearDownFn()
+
+	t.Log("Creating a namespace that will be used by this test")
+	ns := framework.CreateNamespaceOrDie(clientSet, "reflector-fallback-watchlist", t)
+	defer framework.DeleteNamespaceOrDie(clientSet, ns, t)
+
+	t.Logf("Adding 5 secrets to %s namespace", ns.Name)
+	for i := 1; i <= 5; i++ {
+		_, err := clientSet.CoreV1().Secrets(ns.Name).Create(ctx, newSecret(fmt.Sprintf("secret-%d", i)), metav1.CreateOptions{})
+		require.NoError(t, err)
+	}
+
+	t.Log("Creating a secret reflector that will use the WatchList feature to synchronise the store")
+	store, lw := storeAndListWatch()
+	lw.SetClient(ctx, clientSet, ns)
+	target := cache.NewReflector(lw, &v1.Secret{}, store, time.Duration(0))
+	target.UseWatchList = true
+
+	t.Log("Waiting until the secret reflector synchronises to the store (call to the Replace method)")
+	reflectorCtx, reflectorCtxCancel := context.WithCancel(context.Background())
+	defer reflectorCtxCancel()
+	store.SetChanToCloseOnReplace(reflectorCtxCancel)
+	err := target.ListAndWatch(reflectorCtx.Done())
+	require.NoError(t, err)
+
+	t.Log("Verifying if the secret reflector was properly synchronised")
+	verifyStore(t, ctx, clientSet, store, ns)
+
+	t.Log("Starting the second server with the WatchList feature disabled")
+	clientSet2, apiServerTearDownFn2 := startServerWatchListDisabled(t, ctx, etcdOptions)
+	defer apiServerTearDownFn2()
+
+	t.Log("Emptying the store")
+	emptyStore(t, store)
+
+	t.Log("Pointing the ListWatcher used by the secret reflector to the second server (with the WatchList feature disabled)")
+	lw.SetClient(ctx, clientSet2, ns)
+	reflectorCtx, reflectorCtxCancel = context.WithCancel(context.Background())
+	defer reflectorCtxCancel()
+	store.SetChanToCloseOnReplace(reflectorCtxCancel)
+	err = target.ListAndWatch(reflectorCtx.Done())
+	require.NoError(t, err)
+
+	t.Log("Verifying if the secret reflector was properly synchronised")
+	verifyStore(t, ctx, clientSet, store, ns)
```

Comment:
worth trying to sync one more time? (against the server with the feature on)

Reply:
I don't think it would help with anything.